shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java

```diff
@@ -89,6 +91,25 @@ public void createBridgeExecutor(ProcessorDTO processorDTO) {
         }
     }
 
+    private boolean createOrReplace(BridgeExecutor expected, BridgeExecutor existing) {
+        // Does not exist. Create it.
+        if (existing == null) {
+            LOGGER.info("Existing BridgeExecutor does not exist. Creating a new instance.");
+            return true;
+        }
+        // Does exist, but it has already entered the Kubernetes reconciliation loop. Don't update it.
+        // The expected definition will always have READY:Unknown set by BridgeExecutor.Builder.build().
+        // READY:False is set by the first iteration through k8s's reconciliation loop.
+        if (existing.getStatus()
+                .getConditionByType(ConditionTypeConstants.READY)
+                .filter(c -> c.getStatus() == ConditionStatus.False).isPresent()) {
+            LOGGER.info("Existing BridgeExecutor has been marked as READY:False. Skipping.");
+            return false;
+        }
```

Comment:
I'm not sure I get this piece here. If the condition status is not ready **but** the specs are different I think we should update the CRD. In other words I think we should remove these lines, or am I missing anything?

Reply:
Hi @r00ta I'll answer here; but it is probably a general response to your other questions too.

The issue is this: When the `operator` polls the `manager` it receives a `Bridge`/`Processor` that is in the `PREPARED` state. The `operator` then immediately signals to the `manager` that the resources is `PROVISIONING`. If the `operator` dies at this point we are stuck with a resource flagged as `PROVISIONING` that is no longer detected by the `operator` polling the `manager`. The resource never gets completed.

My change now also returns resources from the `manager` that are in `PROVISIONING`. However this has the knock on affect that if Resource X is already being provisioned by the `operator` when a subsequent poll completes the `operator` has two Resource X's to provision; one "in flight" and the other fresh from the subsequent poll.

What do I do!?!

When a new resource is being provisioned the `existing` is `null` and the `expected` has a Status of `Ready:Unknown`. This causes us to invoke request creation and our Controller reconciliation loop starts provisioning the resource. As the first step in the loop I then [set](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/2ef7e7530e8b5cf0ccdebc1dbf66479582cf69c3/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java#L91-L99) the Status to `Ready:False`. 

Should the `operator` die at this point:-
- When it comes back up k8s continues with the reconciliation loop.
- If a poll completes from the `operator`->`manager` the `existing` has a Status `Ready:False` and the check above fails and consequentially we do not signal to k8s to provision the resource (it is already being handled by k8s recovery in the previous bullet point).

Should the operator not die but a subsequent poll completes:
- If a poll completes from the `operator`->`manager` the `existing` has a Status `Ready:False` and the check above fails and consequentially we do not signal to k8s to provision the resource (it is already being handled by k8s recovery in the previous bullet point).

So, I am using the CRD's Status to indicate whether k8s is already handling it.

`Ready:Unknown` -> We need to tell k8s and invoke the controller/reconciliation.
`Ready:False` -> We've already told k8s to handle it.

Reply:
Uhm I still don't get why this piece of code is useful. 
1) If the operator sent the PROVIISONING update but crashed before creating the CRD -> when it starts again it would fetch the resource from the manager and create it
2) If the operator sent the PROVISIONING update and created the CRD it can keep fetching the same resource, but the `createOrReplace` would not have any affect given that the spec is the same. 

Why do we need 
```java 
        if (existing.getStatus()
                .getConditionByType(ConditionTypeConstants.READY)
                .filter(c -> c.getStatus() == ConditionStatus.False).isPresent()) {
            LOGGER.info("Existing BridgeExecutor has been marked as READY:False. Skipping.");
            return false;
        }
```
?

The process should be already resilient if the manager returns the istances in the PROVISIONING state

Reply:
```
In this scenario k8s would continue the reconciliation of (1) and we would also initiate (2) so there would be two reconciliations in process at the same time for the same resource... but please tell me if this is incorrect!!!
```
Yes, this is not correct. The operator sdk guarantees that there is always **exactly** one reconcile loop running for each resource -> the concurrency is already managed by the operator sdk, it basically keeps a queue of triggers and there is only one reconcile loop running for every resource

Reply:
Thanks for the information. That changes _everything_!

Reply:
Thanks for the information. That changes _everything_!